### ISSUE TITLE
Conservation property through layers

### ISSUE THREAD

**Issue by dinhviettoanle** [rachtibat/LRP-eXplains-Transformers#27]
Hey @rachtibat,

Thanks a lot for this library! I'm trying to get the LRP score of each layer on a BERT model by adapting your [example](https://lxt.readthedocs.io/en/latest/feature-visualization.html#latent-feature-attribution). (I'm using the current last commit 6eecc5b, with Python 3.11)

According to the paper, AttnLRP does ensures the conservation property of LRP scores (according to Table A3). However, I've got some value differences when testing it.

# Minimal code to reproduce

```
import torch
from transformers import AutoTokenizer
from lxt.models import bert

def hidden_relevance_hook(module, grad_input, grad_output):
    if isinstance(grad_output, tuple): grad_output = grad_output[0]
    if isinstance(grad_input, tuple): grad_input = grad_input[0]
    module.hidden_relevance = grad_output.detach().cpu()
    module.previous_relevance = grad_input.detach().cpu()

# === Model initialization ===
model = bert.BertForMaskedLM.from_pretrained('bert-base-cased', torch_dtype=torch.bfloat16)
tokenizer = AutoTokenizer.from_pretrained('bert-base-cased')

# === Attach LRP rules and hooks ===
bert.attnlrp.register(model)
model.cls.register_full_backward_hook(hidden_relevance_hook)
model.cls.predictions.decoder.register_full_backward_hook(hidden_relevance_hook)
model.cls.predictions.transform.LayerNorm.register_full_backward_hook(hidden_relevance_hook)

# === Inference and LRP backward pass ===
prompt = f"I have several animals including 5 [MASK] and 3 dogs."

input_ids = tokenizer(prompt, return_tensors="pt", add_special_tokens=True).input_ids.to(model.device)
input_embeds = model.get_input_embeddings()(input_ids)

outputs = model(inputs_embeds=input_embeds.requires_grad_(), output_hidden_states=True) 
max_logits, max_indices = torch.max(outputs.logits[:, 7, :], dim=-1)
print(tokenizer.decode(max_indices)) # >> Output: "cats"

max_logits.backward(max_logits)
```

# Outputs
```
print("Max logits:", max_logits[0].item())
# >> 16.25
print("Output relevance in LM head:", model.cls.hidden_relevance.sum().item())
# >> 16.25
print("Input relevance in LM head:", model.cls.previous_relevance.sum().item())
# >> 18.375
```
This seems to show that the relevance score is modified somewhere in the `(cls): BertOnlyMLMHead(...)` part of the model, at least.

## Possible source of error

By digging into the module itself, the `(decoder): Linear(...)` and all the `Linear(...)` modules seem to ensure this property (with only minor numerical errors I guess):
```
print("Output relevance in LM head decoder:", model.cls.predictions.decoder.hidden_relevance.sum().item())
## >> 16.25
print("Input relevance in LM head decoder:", model.cls.predictions.decoder.previous_relevance.sum().item())
## >> 16.375
```

Though, by moving up in the module, I may managed to narrow it down to the `LayerNormEpsilon` layer which seems to be at the root of the problem:
```
print("Output relevance in LM head LayerNorm:", model.cls.predictions.transform.LayerNorm.hidden_relevance.sum().item())
# >> 16.375
print("Input relevance in LM head LayerNorm:", model.cls.predictions.transform.LayerNorm.previous_relevance.sum().item())
# >> 18.5
```

Is it an implementation issue of the normalization layer or a theoretical misunderstanding from my side dealing with the bias term?
Thanks a lot!

**Comment by rachtibat** [comment]
Hey @dinhviettoanle,

the conservation property does not hold for bias terms in LRP. (This was always the case (: )

> However, (Bach et al., 2015) treats bias terms as
> additional hidden neurons (with an activation value of one
> and a weight that equals the bias value, connected to the
> output) including them into the conservation property (3).
> Consequently, we regard this relevance as preserved, rather
> than lost. Alternatively, to strictly enforce conservation, the
> absorbed relevance score of the bias term can be distributed
> equally among the input variables, or the bias term can be
> excluded completely, as explained in Appendix A.2.1.

In Appendix A.2.1 you get more details.

Since, we model the softmax function with an additional bias term, it also absorbs part of the relevance.

If you want to enforce conservation, simply divide the relevance at the residual stream output by its sum.
I.e. `residual_relevance = residual_relevance / residual_relevance.sum()`

Best

**Comment by dinhviettoanle** [comment]
Okay, thanks for the explanation!
